README.md
# node-red-contrib-mytimeout Readme
Node-red-contrib-mytimeout is a countdown timer that can be trigged by sending it input. It will run until the timer runs out where it will turn off. It can be turned off, stopped or cancelled with the correct input. It can be 'tickled' (restarted) with an on message or any other command not listed below. Meaning, if you send it junk, you have tickled the timer to continue to run. You can dynamically change the timeout and warning values by sending the appropriate Javacript object to the input.

Node-red-contrib-mytimeout can send and receive to the same MQTT topic. If the input matches the previously sent output, the message is ignored to avoid an endless loop.

Moved the default master repos to main repos

## myTimeout
MyTimeout started as hacked timer code I stole from Pete Scargill. Pete's code would start a timer running if you sent it any trigger (tickle the timer). It would continue to run if further triggers were sent before the timeout occurred. A trigger was anything sent to the input of the timer.

My code goes a few steps further. It allows you to tickle the timer, turn off, stop or cancel the timer. It also allows you to send a Javascript object (see below) to override the timer's default values. Allowing you to turn off, to cancel or stop the timer. It still retains some of Pete's behaviors. You can still send almost anything to tickle the timer with the exception of the Javascript objects below which will cause the timer to do specific things.

You can configure the timeout module with the settings for the Safe and Unsafe messages. Safe is sent on the start of the timer, the unsafe message sent when the timeout occurs.

* Name
* Output topic (optional)
* On state payload
* Warning state payload (optional)
* Off state payload
* Countdown (seconds, integer)
* Warning (seconds, integer)

**Please note:** The countdown and warning values must be whole, integer numbers. Using floating point numbers can cause the timer to not send the appropriate off state payload.

All of this is very useful for things like lights that are motion activated. If motion is detected, the Safe payload is sent (default is "on"). The timer can be started when motion is detected and as long as the motion continues there will be no timeout (if motion occurs more often than the default timeout). When the timeout occurs the Unsafe payload is sent (default is "off"). When the timeout warning time (n seconds before the timeout) occurs the warning message is sent on output 1.

If you set the warning seconds to 0 or leave the warning payload empty then the timer will skip sending a warning message and simply do a countdown. Setting the warning time in the Javascript object input to 0 will do the same.

## MyTimeout inputs Preferred method

Like any Node-Red node, MyTimeout accepts the normal Javascript object. The payload attribute and value are expected and two additional attributes are optional. The payload value can be:

* "on" or 1 - which turns on the timer and uses the default settings in the node (and issues an "on" on the output)
* "off" or 0 - which turns off the timer (and issues an "off" on the output)
* "stop" - which stops the timer (and issues a "stop" on the output)
* "cancel" - which cancels the timer (and does not send any output)
* Anything not listed above will be considered an on condition (a kicker if you will)

A minimal Javascript object looks like this:

`
{
   "payload": "on"
}
`

With this message the timer and will run with the default settings in the timer node (timeout is defaulted to 30 seconds and warning is defaulted to 10 seconds).

You can override the timeout and warning times with:

`
{
    "payload": "on",
    "timeout": 3600,
    "warning": 300
}
`

This will restart the time, setting it to timeout in 3600 seconds and issue a warning 300 seconds before the timer turns off.

```
{
    "payload": "on",
    "warning": 0
}
```

Here the timer is turned on, using the default timeout. With the warning set to 0 seconds, the warning message will no longer be sent.

**Please note:** The timeout and warning values must be whole, integer numbers. Using floating point numbers can cause the timer to not send the appropriate off state payload. Both attributes are optional and either can be used to override the default settings.

Other triggers (inputs) are:

```
{
    "payload": "off"
}

{
    "payload": "stop"
}

{
    "payload": "cancel"
}
```

These will stop the timer. See above for the output description.

If the node property 'Warning state payload' is not set in the edit node dialog or the property 'Warning (sec)' is set to 0, then no warning message will be sent.

If the timer is not currently running and a **stop** or **cancel** is sent to the timer, no output will be sent. An **on** will trigger the timer to begin running. If an **off** is sent while the timer is not running, an off message will be sent.

```
{
    "payload": "pause"
}
```

This is a new feature to 4.0.0 where sending a pause will cause the timer to suspend counting down. Sending a second pause or suspend will unpause. @FIXME: This needs more thought

## Special handling

### MQTT

```
mosquitto_pub -t home/automation/timer1 -m '{"payload": "on", "timeout": 600, "warning": 120 }"
```

```
{
    "topic":"home/automation/timer1",
    "payload":"{\"payload\": "\on\", \"timeout\": 600, \"warning\": 120}",
    "qos":0,
    "retain":false,
    "_msgid":"92d84595.fee6f8"
}
```

```
{
    "topic":"home/automation/timer1",
    "payload":"on",
    "timeout":600,
    "warning":120,
    "qos":0,
    "retain":false,
    "_msgid":"92d84595.fee6f8"
}
```
### Warning disabled

```
{
    "topic":"home/automation/timer1",
    "payload":"on",
    "timeout":600,
    "warning":0,
    "qos":0,
    "retain":false,
    "_msgid":"92d84595.fee6f8"
}
```

### Safe message disabled

```
{
    "topic":"home/automation/timer1",
    "payload":"",
    "warning":0,
    "qos":0,
    "retain":false,
    "_msgid":"92d84595.fee6f8"
}
```

## The first output
The first (primary) output sends msg.payload of on, off, warning or stop. An input of cancel, does not send any output.

```
{
    "_msgid":"eed69e39.50a2a",
    "topic":"",
    "payload":"off"
}
```

Here is an off example.

## The second output
This is a new feature in v2.0.0 of node-red-contrib-mytimeout. It is the countdown information.

Example output:
```
{"payload":30,"state":1,"flag":"ticks > 0"}
{"payload":5,"state":2,"flag":"warn >= ticks"}
{"payload": -1, "state": 0, "flag": "stop"}
{"payload": 0, "state": 0, "flag": "off"}
{"payload": -1, "state": 0, "flag": "cancel"}
{"payload": -1, "state": 0, "flag": "unknown"}
```

The "payload" is the number of seconds left on the timer, if 0 that means it is off, -1 means it was stopped, cancelled or the unknown has occurred.

Also I've added a second output which sends out the time left on the timer.

```
{"payload":30,"state":1,"flag":"ticks > 0"}
or
{"payload":10,"state":2,"flag":"warn >= ticks"}
or
{"payload":0,"state":0,"flag":"off"}
or
{"payload":-1,"state":0,"flag":"stop"}
or
{"payload":-1,"state":0,"flag":"cancel"}
or
{"payload": -1, "state": 0, "flag": "unknown"}
```

Then the **off** is sent or timeout occurs, the output 2 message is the same.
The **unknown** should never occur but is left in in case something breaks.

The states are:
*  0 = off, stopped or cancel (timer not running)
*  1 = on (timer is running)
*  2 = warning (timer is running and in the warning period)

If you find any other state, open an issue.
[node-red-contrib-mytimeout Issues](https://github.com/linuxha/node-red-contrib-mytimeout/issues)

Actual output:
```
{
    "payload":4,
    "state":2,
    "flag":"warn >= ticks",
    "_msgid":"17b809b8.85c0d6"
}
```

### Payload
* **on** - turns on the timer. The addition fields *timeout* and *warning* are optional and allow the user to change the defaults. Both field values are integers and are in seconds. This payload will cause the timer to send a *Timer on payload* message in the msg.payload output

CHANGELOG.md
## 3.2.3 (February 13, 2022)
  - Added test suites, not 100% confident all the tests are correct yet
## 3.2.2 (February 12, 2022)
  - Gave up even trying to fix the pre-3.2.0 part of the Change log.
  - Removed unneeded dependency in package.json
  - Another update to fix 'score' issue
  - Added Engines and dependencies to package.json
  - Updated mytimeout.js version to reflect added examples
  - Updated package version to reflect added examples
  - Added an example flow to the pacakge
## 3.2.1 (February 11, 2022)
  - Oops, missed a few updates
  - Minor change to add examples to the package
## 3.2.0 (October 22, 2019)
  - Minor cleanup of comments and updated major version to 3.2.0 for the No On Feature
  - Updated major for the no on feature
  - Merge pull request #25 from linuxha/feature/no_on_payload
  - Corrected a missed editor save
  - Update script to use local variables in the functions and added more tests.
  - Added No On message feature, fixed attibutes, warning, and Float bugs.
  - Added the test flow used by the simple.sh script.
  - Added a bash script for manual pair-wise testing.
  - Feature request: adjusted code to allow a blank safe msg to not be sent.
  - Feature request: removed default "on" and "require" for safe message
  - Bram Hoefnagel (brammh) fixed missing topic in out message
  - Bumped minor +1
  - Minor corrections to the README
  - Merge pull request #20 from brammh/Bugfix-#16---topic-name-included-in-control-messages
  - Update mytimeout.js
  - Change to version 3.0.2
  - As per Colin Law I've changed the wording to refer to Javascript object.
  - Added parseInt to the timeout and warning values.
  - Updated version for node.warn fix
  - Fixed node.warn error message
  - Updated to fix output2 to contian the correct msg.payload for time remaining.
  - Updated version to 2.2.3 for force integer input
  - Update documentation to indicate that time must be in integer seconds
  - Updated to handle incoming exceptions as on payloads
  - Incremented version for deploy fix
  - Incremented the minor version for deploy fix
  - Removed the required on the Countdown, this caused the deploy errors
  - Merge pull request #8 from linuxha/new-features-mytimeout
  - Added new feature to ignore case on input match.
  - Updated package.json to 2.1.9 for bug fix
  - Merge pull request #7 from linuxha/bug-fix-mytimeout
  - Modified newMsg to handle msg.payload { payload, timeout, warning } properly
  - Pushed minor version increment to unstick npm publish
  - Updated with debug option, v2.1.7
## 2.1.5 (August 3 2018)
  - Added a debug flag to the node
  - Updated the documentation and added a note on a work around
  - Added note to a work around for a problem with multiple timers
  - Don't trust the test.js, it passes when it shouldn't
  - Enhanced and massaged the wording in the README
  - Added support for turning on/off debug logging in the node
  - Update version to 2.1.4 and change test to use mocha
  - Updated, as per node-red-node-test-helper Readme
  - Documentation clean up and reorganizing
  - Incremented minor version
  - Added a link to the sample flow
  - More documentation added
  - More documentation added
  - Major update of the message, timeout and warning handling.
  - Updated the minor version
  - Updated the version change (minor)
  - Updated some more documentation
  - updated some more documentation
  - Updated to version 2.0.0
  - Updated documentation to reflect change to 2.0.0 and some of the differences
  - Major changes, added feature and changed the JSON input timer to timeout.
  - Update comments, documentation and added 1 more output for seconds remaining
  - Added comment about needing to work on the state engine
  - Incremented minor version number
  - incremented version so I can publish changes
  - Reformatted the html and added a limit vaule
  - Reformatted the html and added a limit vaule
  - Updated keywords to include more keywords like node-red
  - 1.0.13
  - Moved under cherry instead of scargill
  - 1.0.12

mytimeout.js
// Author: Neil Cherry <ncherry@linuxha.com>

// I'd be much better off with a state machine
// https://www.smashingmagazine.com/2018/01/rise-state-machines/

// the on/off trigger string below whould be user configurable
//
// Inputs:
//  msg.payload = {
//    "payload": "on",
//    "timeout": 3600,
//    "warning": 300
//  }
//
// The above can be used to change the timeout and warning
//
// or
//
//  {
//    "payload": "on"
//  }
//
//  {
//    "payload": <Don't care>
//  }
//
// The above gets treated as an on condition
//
//  {
//    "payload": "off"
//  }
//
//  {
//    "payload": "off"
//  }
//
//  {
//    "payload": "stop"
//  }
//
//  {
//    "payload": "cancel"
//  }
//
//  {
//    "payload": "tick"
//  }

// Initial is stop
// STOP => RUN => STOP (via timeout)

module.exports = function(RED) {
    "use strict";

    /*
      node-input-name
      node-input-outtopic
      node-input-outsafe
      node-input-outwarning
      node-input-warning
      node-input-timer
      node-input-limit
      node-input-repeat
      node-input-again
      node-input-atStart
    */
    var countdownNode = function(n) {
        // Local variables
        var node        = this;
        
        var state       = 'stop';
        var wflag       = false;
        var ticks       = -1;           //
        var lastPayload = Date.now();;  // 
        var pauseValue  = null;         // Feature: pause
        var tixID       = null;         // Feature: pause

        var timeout     = parseInt(n.timer||30);    //
        var warn        = parseInt(n.warning||10);  //

        var ignoreCase  = '';

        var line        = {};
        var version     = '4.0.0'; //

        RED.nodes.createNode(this, n);

        // GUI variables
        node.timer     = parseInt(n.timer||30); // Need to build checking into the html file
        node.state     = 'stop';                // For now it's a string, later an object?
        /*
          This is the Edit dialog for this node
          node properties
          Name                      Countdown timer
          Output Topic              topic <- if blank no output topic
          Timer On payload          on
          Warning state payload     Warning
          Timer Off payload         off
          Countdown (secs)          30
          Warning (secs)            5

          Rate Limit (msg/secs)     30
          [ ] Repeat message every second <- This doesn't seem like a good idea (???)
          [ ] Auto-restart when timed out
          [ ] Run at start

        */
        node.name      = n.name;               // node-input-name       - Name
        node.warning   = parseInt(n.warning||5);// node-input-warning   - time in seconds (?)
        node.topic     = n.outtopic;           // node-input-outtopic   - Output topic
        node.outsafe   = n.outsafe;            // node-input-outsafe    - Timer on payload
        node.outwarn   = n.outwarning;         // node-input-outwarning - Warning state payload
        node.outunsafe = n.outunsafe || "off"; // node-input-outunsafe  - Timer off payload
        //                                     // node-input-warning    - warning seconds
        //                                     // node-input-timer      - countdown seconds
        //                                     // node-input-repeat     - Rate limit seconds
        node.repeat    = n.repeat;             // node-input-repeat     - Repeat message every second
        node.again     = n.again;              // node-input-again      - Auto restart when timed out
        node.atStart   = n.atStart;            // node-input-atStart    - Run at start

        //
        function ndebug(s){
            // debug is either true or false
            if(n.ndebug) {      // only if true do we debug
                node.log(s); // Can't use the obj this here, used obj node
            }
        }

        if(n.ignoreCase) {      // Only if true do we ignore case
            ignoreCase = 'i';
        }

        // this.log("Debug = (" + n.ndebug + ")"); // Can't use the obj node here, use obj this
        this.status({
            fill  : "red",
            shape : "dot",
            text  : "Stopped: init"
        });

        // =====================================================================
        // There can be:
        // payload = "on" or 1 or "1" or anthing else except off/stop/cancel/debug
        function on(msg) {
            // My intention is to move all the calculations for
            // these variables to here. At the moment they're all over
            // the place and confusing
            if(n.ndebug) {
                ndebug("    msg:        " + JSON.stringify(msg));
                ndebug("    msg:        " + typeof(msg));
                ndebug("    node.timer: " + node.timer);
                ndebug("    node.warning: " + node.warning);
                try {
                    ndebug("    msg.timeout:" + msg.timeout);
                    ndebug("    msg.warning:" + msg.warning);
                } catch(e) {
                    ndebug("    msg.timeout:undefined");
                    ndebug("    msg.warning:undefined");
                }
                ndebug("    timeout:    " + timeout);
                ndebug("    warn:       " + warn);
                ndebug("    ticks:      " + ticks);
            }
            //
            // There are 3 sets of variables
            // default values (node.timer, node.warning)
            // passed values  (timeout, warn - if any)
            // running values (ticks)
            //
            timeout = msg.timeout||timeout||node.timer;
            timeout = parseInt(timeout);                // ncherry@linuxha.com parseInt()
            if(msg.warning != 0) {
                warn = msg.warning||warn||node.warning;
            } else {
                warn = 0;
            }
            warn    = parseInt(warn);                   // ncherry@linuxha.com parseInt()

            ticks = timeout;

            if(n.ndebug) {
                ndebug("");
                ndebug("    node.timer: " + node.timer);
                ndebug("    node.warning: " + node.warning);
                ndebug("    timeout:    " + timeout);
                ndebug("    warn:       " + warn);
                ndebug("    ticks:      " + ticks);

                ndebug("Count timer on");
            }
            node.status({
                fill  : "green",
                shape : "dot",
                text  : "Running: " + ticks // provide a visual countdown
            });

            // @Feature: no on msg
            if(msg.payload) {
                msg.payload = node.outsafe;
            }
            msg.topic   = node.topic;
            lastPayload = msg.payload;
            ndebug("Send Green: " + lastPayload || "''");
            if(msg.payload) {                             // @FEATURE: no on msg
                node.send([msg, null]);
            }

            state = 'run';
            if (tixID !== null) {
              clearInterval(tixID);
                tixID = null;
            }
            startInterval();
            
            wflag = false;      // rest the warning flag
        } // on(msg)

        function off() {
            ndebug("off!");
            ticks = -1;
            stop('off');
        } // off()

        // I'm about to make this a bit complicated
        // In: off    out: off
        // In: stop   out: stop
        // In: cancel out: (nothing)
        function stop(s) {
            // The states are only called from the node.on("input", ...)
            // That's where I define the line obj after I massage the incoming
            // message
            var msg = line;     // This is a biy of a risk, I'll leave it for now
            // if the main input routine calls this function, it will pass an
            // object (the input msg) which we don't care about really
            if(typeof(s) !== "string") {
                ndebug("Empty stop");
                s = 'stop';
            }

            ndebug(s + "! A");
            ticks = 0;

            node.status({
                fill  : "red",
                shape : "dot",
                text  : "Stopped: " + s // provide a visual countdown
            });

            // Stop or off can send, not cancel
            switch(s) {
                case 'stop':
                    msg.payload = "stop";
                    lastPayload = msg.payload;
                    ndebug("Send Red: " + lastPayload);

                    var tremain = {"payload": -1, "state": 0, "flag": "stop"};
                    node.send([msg, tremain]);
                    break;

                case 'off':
                    msg.payload = node.outunsafe;
                    msg.topic = node.topic;
                    lastPayload = msg.payload;
                    ndebug("Send red: " + lastPayload);

                    var tremain = { "payload": 0, "state": 0, "flag": "off"};
                    node.send([msg, tremain]);
                    break;

                case 'cancel':
                    ndebug("Send red: null");
                    var tremain = { "payload": -1, "state": 0, "flag": "cancel"};
                    lastPayload = Date.now();
                    node.send([null, tremain]);
                    break;

                default:
                    ndebug("Send red: ???");
                    var tremain = { "payload": -1, "state": 0, "flag": "unknown"};
                    lastPayload = Date.now();
                    node.send([null, tremain]);
                    break;
            }

            if (tixID !== null) {
                clearInterval(tixID);
                tixID = null;
            }
            state = 'stop';
            ticks = -1;
            timeout = parseInt(node.timer);
            warn    = parseInt(node.warning);

            ndebug('=[ fini ]=======================================================================');
        }

        function cancel() {
            ndebug("cancel!");
            stop('cancel');
            ticks = -1;
            pauseValue = null;
        }

        /*
        ** What is being sent here as the message
        */
        function pause(susp) {
            var suspend = susp.payload == "suspend" ? true : false;
            ndebug("pause function!");
            pauseValue = [ticks, node.warnT];

            timeout = ticks;
            var msg = line;
            node.status({
                fill  : "grey",
                shape : "dot",
                text  : `Paused: ${ticks}` // provide a visual countdown
            });

            msg.payload = "pause";
            lastPayload = msg.payload;

            state = 'pause';
            if (tixID !== null && !suspend) {
                var tremain = { "payload": ticks, "state": 0, "flag": "pause"};
                clearInterval(tixID);
                tixID = null;
            }
            node.send([msg, tremain]);
        }

        function unpause(msg) {
            ndebug("unpause!");

            var pausemsg = {
                timeout: pauseValue[0],
                warning: pauseValue[1]
            }
            on(pausemsg);
        }

        function doNothing() {
            ndebug("doNothing!");
            state = 'stop';
            ticks = -1;
            pauseValue  = null; // I think this needs to be here
        }

        // @TODO: This should return the original msg with as few changes as possible
        function newMsg(msg) {
            // Pretty much a deep clone but I don't think it
            // will copy things like sockets
            //nMsg.payload = JSON.parse(JSON.stringify(msg));
            var nMsg = Object.assign(msg, {}); // create a copy perhaps use this: JSON.parse(JSON.stringify(msg));

            // Okay now I need to figure out the format
            // is it?
            // A: { "payload": "on", "timeout": 30, ... }
            // or
            // B: { "payload": { "payload": "on", "timeout": 30, ... } }
            // or
            // C: { "payload": "{ \"payload\": \"on\", \"timeout\": 30, \"extra\":\"inside\" }", "other":"other", "extra": "outside" }
            switch(typeof(msg.payload)) {
                case "string":
                    ndebug("Str msg  = " + JSON.stringify(msg));
                    if(/.*"payload".*/.test(msg.payload)) {
                        ndebug(" msg.payload  = " + JSON.stringify(msg.payload));

                        // string contain payload, convery to JSON
                        var t = newMsg(JSON.parse(msg.payload));
                        nMsg = Object.assign(msg, t); // Perhaps I should use nMsg instead msg (and copy msg to nMsg above)?
                    }

                    //
                    try{
                        nMsg.payload = nMsg.payload.toLowerCase();
                    } catch(e) {
                        nMsg.payload = nMsg.payload.toString().toLowerCase();
                    }
                    break;

                case "number":
                    ndebug("Num msg  = " + JSON.stringify(msg));
                    nMsg.payload = msg.payload.toString();
                    break;

                case "object":
                    var m = msg.payload;
                    ndebug("Obj msg  = " + JSON.stringify(m));
                    t    = JSON.parse(JSON.stringify(m));
                    ndebug("obj m    = " + JSON.stringify(m));
                    nMsg = Object.assign(msg, t);
                    ndebug("obj t    = " + JSON.stringify(t));
                    ndebug("obj nmsg = " + JSON.stringify(nMsg));
                    break;

                default:
                    ndebug("??? msg  = " + JSON.stringify(msg));
                    ndebug("??? msg  = " + typeof(msg.payload));
                    nMsg.payload = "";
                    break;
            }
            ndebug("RTN msg  = " + JSON.stringify(nMsg));
            
            return nMsg;
        }

        // Leave this here, need the functions ref from above
        // states[state][line.payload](line); // function call to line.payload(line)
        // <state>:<function>
        // on: on
        // pause: pause
        // suspend: pause
        // maybe wait/cont
        // definitely pause/unpause but this seems to come with a true false (???)
        // definitely suspend/restore
        var states = {
            // Not sure if this is what I want in the long run but this is good for now
            stop: { 0: off, on: on, pause: on, suspend: off, off: off, stop: doNothing, cancel: doNothing },
            pause: { 0: off, on: on, pause: unpause, suspend: unpause, off: off, stop: stop, cancel: cancel },
            run:  { 0: off, on: on, pause: pause, suspend: pause, off: off, stop: stop, cancel: cancel }
        };

        // -------------------------------------------------------------------------------
        // Commands
        // TIX
        node.on("TIX", function(inMsg) { // old has , state)
            lastPayload = Date.now();
            var msg = {};

            if(ticks > 0) {
                // A blank outwarning means don't send
                if((warn >= ticks) && (node.outwarn !== "")) {
                    // a warn of 0 seconds also means don't send
                    if (warn) {
                        // Timer at warning
                        node.status({
                            fill  : "yellow",
                            shape : "dot",
                            text  : "Warning: " + ticks // provide a visual countdown
                        });

                        if(!wflag) {
                            msg = Object.assign(msg, line); // Copy the original attributes into the new msg

                            msg.payload = node.outwarn;
                            msg.topic   = node.topic;
                            lastPayload = msg.payload;
                            ndebug("Send Yellow: " + lastPayload);
                            node.send([msg, null]);
                            wflag = true;
                        }
                    } // warn if there's a warn message

                    var tremain = { "payload": ticks, "state": 2, "flag": "warn >= ticks"};
                    node.send([null, tremain]);
                } else {
                    // HOW does this get sent out?
                    node.status({
                        fill  : "green",
                        shape : "dot",
                        text  : "Running: " + ticks // provide a visual countdown
                    });

                    var tremain = { "payload": ticks, "state": 1, "flag": "ticks > 0"};
                    node.send([null, tremain]);
                }

                state !== "pause" ? ticks-- : '';
            } else if(ticks == 0) {
                ndebug("ticks == 0");
                stop("off");

                //var tremain = { "payload": 0, "state": 0, "flag": "ticks == 0"}};
                //node.send([null, tremain]);

                ticks = -1;
            } else {
                // Do nothing
            }
        }); // node.on("TIX", function {});

        // Stop         (initial state at start up and when not running)
        // On           (timer reset to default value and running, on sent)
        // Pause        (timer retains value, return to Stop)
        // Off          (timer off, off sent, return to Stop)
        // Cancel       (timer off, nothing sent, return to Stop)
        // Warning      (timer still on, warning sent)
        // Timeout      (timer off, off sent, return to Stop
        node.on( "input", function(inMsg) {
            // inMsg = {"topic":"home/test/countdown-in-b","payload":"{ \"payload\":\"on\",\"timeout\":6,\"warning\":3}","qos":0,"retain":false,"_msgid":"10ea6e2f.68fb32"}
            // inMsg = {"topic":"home/test/countdown-in-b","payload":"on","qos":0,"retain":false,"_msgid":"fd875a01.526a68"}
            if(n.ndebug) {
                ndebug('=[ input ]======================================================================');
                ndebug('1 node.input("input");');
                ndebug("1 inMsg = " + JSON.stringify(inMsg));
                ndebug("1 State = " + state);
                ndebug("1 timeout = " + node.timer + " - node.timer");
                ndebug("1 timeout = " + timeout + " - timeout");
                ndebug("1 warning = " + node.warning);
            }
            // =================================================================
            // First we need to drop any message than matches the last message
            // sent. This will keep us from getting into an infinite loop.
            // =================================================================

            // From here to the try { } catch {} we should only need to get the
            // inMsg's payload into line.payload (could be a string or object)


            // This is taken from myTimeout.js (node-red-contrib-mytimeout)
            // It's purposed is to stop from repeating the same message from
            // being resent (an endless loop). For some reason this is on the
            // isString() check ... hmmm
            // =================================================================
// ================================================================================
            // =========================================================
            // Okay here's what I expect, user sends something (anything)
            // node send node.outsafe,
            // if node sees the exact string node.outstring just sent
            // then drop it
            // If the timer goes off then the lastPayload should be cleared

            // We only send a simple message ('on', 'pause', 'off', 'stop' or 'cancel')
            var regex = new RegExp('^' + lastPayload + '$', ignoreCase);
            //if(lastPayload === inMsg.payload) {
            if(regex.test(inMsg.payload)) {
                // So it's the same message as what was previously sent
                ndebug("4 TO: In == Out match, skip (" + lastPayload + "/" + inMsg.payload + ")");
                ndebug('=[ Skip ]=======================================================================');
                lastPayload = Date.now();
                return ; //
            }

            try {
                line = newMsg(inMsg);
            } catch (err) {
                // Basically treat the unknown payload as an on (i.e. anything can
                // tickle the timer as long as it's not off, 0, stop or cancel)
                node.warn("L528: newMsg(inMsg): " + err + " <" + JSON.stringify(inMsg) + ">");
            }
            ndebug("line = " + JSON.stringify(line));

// ================================================================================
            var s = "nada";
            
            try {
                states[state][line.payload](line);
            } catch(err) {
                // =============================================================
                // Anything that is not an existing state/function is treated as
                // an on request.
                // =============================================================
                ndebug("state = <" + line.payload + ">" + " (this is not an error - treated as 'on')");
                ndebug("states catch: " + err + " (" + ticks + "/" + warn + " - this is not an error)");
                // If it's not an existing state then treat it as an on
                // that way anthing can be used as a kicker to keep the timer
                // running
                on(line);
            }

            /* This is handled in newMsg(msg) above
            try {
                if(typeof(line.payload) == 'string') {
                    s = ("1 states catch: line \"" + line.payload);
                    states[state][line.payload](line);
                } else {
                    // Hopefully I've converted all the input to a string and
                    // to lower case.
                    s = ("2 states catch: line  \"" + line.payload.payload);
                    states[state][line.payload.payload](line);
                }
            } catch(err) {
                // =============================================================
                // Anything that is not an existing state/function is treated as
                // an on request.
                // =============================================================
                ndebug(s  + " \" (this is not an error)");
                ndebug("states catch: " + err + "(" + ticks + "/" + warn + " - this is not an error)");
                // If it's not an existing state then treat it as an on
                // that way anthing can be used as a kicker to keep the timer
                // running
                on(line);
            } /* */
// ================================================================================
        }); // node.on("input", ... )

        // Once the node is instantiated this keeps running
        // I'd like to change this to run when only needed
        // Once the node is instantiated this only runs on call on()
        function startInterval() {
          var msg = { payload:'TIX', topic:""};
          node.emit("TIX", msg, state);
          tixID = setInterval(function() {
              node.emit("TIX", msg, state);
          }, 1000); // trigger every 1 sec
        }
        node.on("close", function() {
            if (tixID !== null) {
                clearInterval(tixID);
            }
        });

    } // function myTimeoutNode(n);
    RED.nodes.registerType("mytimeout", countdownNode);
} // module.exports
